### ISSUE TITLE
1.4.3 distortion after glitch on Huawei Honor 8

### ISSUE THREAD

**Issue by serhii-k** [google/oboe#1047]
Android version(s): 8.0
Android device(s): Huawei honor 8
Oboe version: 1.4.3
App name used for testing: OboeTester
 
**Short description**
The resampler in 1.4.3 has some bug. After pausing-starting (1 or several times) it starts to distort the output.

I have previously used Oboe 1.2, and recently noticed that the new Oboe can convert the sample rate. So I've tried this feature with Oboe 1.4.3.
1) with Oboe 1.4.3 release downloaded here:
https://github.com/google/oboe/archive/1.4.3.zip
My app outputs 44100 Hz audio data, which is converted to the device 48000 Hz.
When I pause the audio stream (requestPause), and resume it (requestStart) just once, or sometimes 2-3 times, audio becomes useless. The audio output is just some high quiet whistle.
2) if I clone the latest master Oboe branch, after pause-resume the audio is recognizable but it becomes distorted.

**Steps to reproduce**
This issue can be reproduced with OboeTester -> Test Output.
Settings:
API: OpenSL ES (my phone has Android 8.0)
SRate: 44100 (my phone's preferred sample rate is 48000)
Channels: 2
Perf: LOW_LATENCY
Convert: [V] Chan, [V] Format, SRC: Medium
1) OPEN
2) START
3) press PAUSE-START (with or without any delay in between) once or a couple of times

Hm, after trying to make a recording, I've noticed that with OboeTester I had to cold-restart the app to replicate this issue.

**Expected behavior**
No distortion in the output signal after resuming.

**Actual behavior**
After resuming, you will notice the output becomes distorted.

**Device**
I assume this issue is not device specific. As I can remember I had it even on some emulator.
I guess it can be something related to the latest stored state of the resampler.

I can reproduce this issue on Huawei honor 8, Android 8.0.

**Any additional context**

If applicable, please attach a recording of the sound.

Here is a recording:

[untitled.wav.zip](https://github.com/google/oboe/files/5408637/untitled.wav.zip)

**Comment by philburk** [comment]
Thank you for the clear report and WAV file.

@serhii-k - I have some questions:
1. Was the WAV file recorded with Oboe 1.4.3 or the latest from GitHub?
1. If you close the stream, change to Low or High quality SRC, then retest, does the distortion change?
1. How often does this happen? Can you reproduce it reliably?
1. Have you tried this on different devices or Android versions?

I cannot reproduce this on Pixel 3a but I think I heard it happen once.
I also tried on an emulated Nexus running NYC (Android 7.0).

From the WAV file I can see a major cycle of around 400 frames and a minor cycle of around 10 frames.

I did an experiment with initializing mCoefficientCursor to 5 in PolyphaseResampler.h.
That results in a corruption very different than  the reported sound.

Note that when you are hitting START and PAUSE that it is very easy to hit the "Workload" slider. That can cause under-runs and distortion. But the WAV file in comment 1 does not sound like that.

**Comment by serhii-k** [comment]
Hi @philburk,
Thank you for your reply!
1. The sent previously WAV files was recorded with the latest Oboe from GitHub.
2. Yes, it happens for Low and High quality as well (after closing the stream).
3. Oh, this is a good question. Most of the time I can reproduce it absolutely reliably, it happens if I pause-start the stream 1, 2 or 3 times. But if that doesn't happen after 5 resumes, it seems it's no longer reproducible (until closing and opening a new stream. That might happen, but rarely. Most often it is possible to cause the distortion on a stream.
However, that was a time that it was actually hard to reproduce this issue. I had to cold-restart OboeTester to hear it again.
4. Nope, sorry. I only also have an old phone with Android SDK 16, and it can't run OboeTester. However, in my app I tried to reproduce the conversion from 48000 to its native 44100, and the resampler worked perfectly.

After some investigation, I'm starting to realize that this issue might be not related to the resampler. But I will return to this point later.

Here are some new recordings:
https://drive.google.com/file/d/1Y3kCZba-bIg1RFvscY5LdA48R8Dcxr4b/view?usp=sharing

I initially recorded them (with Audacity) by placing my phone next to my computer. But later on I had realized that it is possible to record a device video with the system audio!

The distorted audio seems to fade out, but it actually remains constant.

1) Files inside the '1.4.3-release' folder were recorded with the released Oboe 1.4.3. In this case, after resuming the audio output often becomes just some very high wave. No distrotion in this case, just some weird high pitch. If you have pets at home, please hold them far away from your computer :)

2) Files inside the 'git' folder were recorded with the latest git version. There are 7 attempts with the Medium quality SRC, they were produced one after another. So this distortion happens really often. And it is not the same each time.

Then, I started making the videos. After watching the first one, I was really surprised. As this issue occured very rarely! It seems it becomes shy when it's recorded. So I have recorded one more. And again it appeared only once.

FYI the cycles when this issue appears starting at:
video1.mpg
00:33 (Medium) and 02:18 (Low)
video2.mpg
00:17 (Medium)
It happens with the High quality as well, but that was not recorded.

---

One special thing I've noticed is this glitch that appears exactly at 02:42 in video2.mpg. I hear it really often when after pause-start the stream remains not distorted. And it seems, such glitches initiate the distortion if it occurs.

However, here is an interesting part. To try this new sample rate conversion feature, I had to open the LowLatency stream. Previously, it was not enabled in my app (well, some time ago I applied the LowLatency + Exclusive combination, but several people started complaining, so I had to remove them).

So within OboeTester -> Test Output I've tried to test the following:
API: OpenSL ES (my phone has Android 8.0)
SRate: 48000 (native here)
Channels: 2

And I have recorded a new video with Perf: LOW_LATENCY and Perf: NONE. Here it is:
https://drive.google.com/file/d/1F-4L9YIP-r19cn17fAg3VX3HifBdUvKl/view?usp=sharing

Glitches appear on this video at 00:05 and 00:46, when the LowLatency mode is enabled.

So, maybe this issue is somehow related to the LowLatency mode on my phone... And there is no issue with the reampler.

Sorry for such long response, and so many recordings... Just to provide the fullest details I'm also adding the log after pressing the "OPEN" stream (configuration: OpenSL ES, 48000, 2 channels, LOW_LATENCY).

2020-10-22 15:57:59.916 15139-15139/com.google.sample.oboe.manualtest V/AudioManager: playSoundEffect   effectType: 02020-10-22 15:57:59.917 15139-15139/com.google.sample.oboe.manualtest D/TestOboe: closeAudio() already closing2020-10-22 15:57:59.918 15139-15139/com.google.sample.oboe.manualtest D/OboeTester: OboeAudioStream_openNative: sampleRate = 48000, framesPerBurst = 9602020-10-22 15:57:59.918 15139-15139/com.google.sample.oboe.manualtest D/OboeAudio: ActivityContext::open() set callback to use oboeCallbackProxy, callback size = 02020-10-22 15:57:59.919 15139-15139/com.google.sample.oboe.manualtest I/OboeAudio: openStream() OUTPUT -------- OboeVersion1.4.3 --------2020-10-22 15:57:59.920 15139-15139/com.google.sample.oboe.manualtest D/OboeAudio: AAudioLoader():  dlopen(libaaudio.so) returned 0x5a92fe5e72cc0c592020-10-22 15:57:59.921 15139-15139/com.google.sample.oboe.manualtest W/OboeAudio: SharingMode [AudioStreamBuilder::setSharingMode()] is not supported on OpenSLES streams.2020-10-22 15:57:59.921 15139-15139/com.google.sample.oboe.manualtest I/OboeAudio: AudioStreamOpenSLES::open() chans=2, rate=480002020-10-22 15:57:59.923 15139-15139/com.google.sample.oboe.manualtest I/libOpenSLES: Emulating old channel mask behavior (ignoring positional mask 0x3, using default mask 0x3 based on channel count of 2)2020-10-22 15:57:59.929 15139-15139/com.google.sample.oboe.manualtest W/AudioTrack: AUDIO_OUTPUT_FLAG_FAST denied by server; frameCount 1924 -> 19242020-10-22 15:57:59.932 15139-15139/com.google.sample.oboe.manualtest D/AudioTrack: OFFLOAD 0,mNotificationFrames -2,mStreamType =3,mOriginalSampleRate 48000,mAfSampleRate 48000,mTransfer 12020-10-22 15:57:59.932 15139-15139/com.google.sample.oboe.manualtest D/AudioTrack: Client defaulted notificationFrames to 480 for frameCount 19242020-10-22 15:57:59.934 15139-15139/com.google.sample.oboe.manualtest W/libOpenSLES: Leaving AndroidConfiguration::GetConfiguration (SL_RESULT_FEATURE_UNSUPPORTED)2020-10-22 15:57:59.935 15139-15139/com.google.sample.oboe.manualtest D/OboeAudio: AudioStreamOpenSLES:configureBufferSizes(48000) initial mFramesPerBurst = 0, mFramesPerCallback = 02020-10-22 15:57:59.935 15139-15139/com.google.sample.oboe.manualtest D/OboeAudio: AudioStreamOpenSLES:configureBufferSizes() NOT low latency, set mFramesPerBurst = 9602020-10-22 15:57:59.935 15139-15139/com.google.sample.oboe.manualtest D/OboeAudio: AudioStreamOpenSLES:configureBufferSizes(48000) final mFramesPerBurst = 960, mFramesPerCallback = 9602020-10-22 15:57:59.935 15139-15139/com.google.sample.oboe.manualtest D/OboeTester: OboeAudioStream_getNativeApi got 1

**Comment by philburk** [comment]
I changed the name of the bug to reflect the probably relation to glitching.

I suspect that OpenSL ES is not recovering from the buffer underflow.

You can this by starting an OpenSL ES output stream by increasing the "Workload" using the fader until you hear constant glitching then quickly move the Workload back to zero.  Does it stay corrupted?

We may be able to improve this by increasing the number of Buffers that OpenSL ES uses. See #744 

As an experiment, try setting kBufferQueueLength = 8.

https://github.com/google/oboe/blob/395f3d6ac25c2b069d53451b89dff4aa96d26eb8/src/opensles/AudioStreamOpenSLES.h#L33

**Comment by serhii-k** [comment]
Here is the result:
1. I've opened, and started an OpenSL ES stream (it became 48000, 2 channels, float).
2. I've gradually increased the "Workload" til glitching appeared, the quickly moved it back to zero.
3. The audio no longer has glitches caused by the "Workload".
4. I've tried to do pause-start, and after 2-3 attempts I still hear the recorded previously initial glitch on start.

Then, I've set kBufferQueueLength = 8 inside the AudioStreamOpenSLES.h.
Everything remains same.

@philburk, I'm really sorry if I'm bothering you with this issue. For me it doesn't have any priority. So it may be not as important as other existing issues, and may be postponed for the future.

**Comment by philburk** [comment]
It is not a bother. We really appreciate these reports. There are hundreds of Android phone models out there and we only have a few for testing.

It is looking like the problem may be related to an Issue in the Huawei HAL or DSP, when restarting a stream. I cannot reproduce it on my devices using OboeTester. Hopefully Huawei has fixed this in later versions.

Unfortunately, we are not able to work closely with Huawei because of trade restrictions. So there is not much I can do at this point.

You could try experimenting with different OpenSL ES buffer sizes using this global before opening OpenSL ES:

    oboe::DefaultStreamValues::FramesPerBurst = (int32_t) framesPerBurst;

If it helps let me know. If not then we can close this.

**Comment by serhii-k** [comment]
Hm, yes it really looks like some hardware issue... I've tried to play with oboe::DefaultStreamValues::FramesPerBurst. That gave very promising results! But the ending distortion still occurs rarely.

I've played with my app (not OboeTester). Previously the oboe::DefaultStreamValues::FramesPerBurst was initialized as AudioManager.PROPERTY_OUTPUT_FRAMES_PER_BUFFER (accordingly to the Getting Started tutorial). Which on Huawei honor 8 is **960**.

Also I want to mention that I tested the released Oboe 1.4.3, not the latest git version. So in this case, after pause-start the broken audio output sounds like some high pitch.

Here is one important observation. I started to test this issue by running a debug build. And it was really hard to reproduce (still possible, but very rarely). (Interestingly, with OboeTester the issue could be easily reproduced with a debug build).

Afterwards, I worked with a profile build. And the distortion could be reproduced very easily for:
oboe::DefaultStreamValues::FramesPerBurst = 960
It happened on the second-third pause-start. But on a debug build, it was not that easy.

Another important thing to mention. For some reason, Oboe can't start a LowLatency stream on my phone (I have previously sent the logcat example). So the opened stream has getPerformanceMode() == None. But it still uses the flowgraph:
I/OboeAudio(27221): configure() flowgraph converts channels: 2 to 2, format: 2 to 2, rate: 44100 to 48000, qual = 3

(This is why initially I thought it might be an issue with the resampler.)

One more strange thing is that after opening a stream, its getFramesPerCallback() returns 0. For a stream with an initially requested PerformanceMode None, this method returns the same value as getFramesPerBurst().

Now experiments with the oboe::DefaultStreamValues::FramesPerBurst.

1) I set it to 128, after opening a stream getFramesPerBurst() returns 1024.
It was a big surprise, but at first I couldn't reproduce the distortion. However, after some time it happened. It was really much harder to achieve.

2) I reverted back to 960. The issue is reproducible SO easily.

3) 800 -> after opening a stream, getFramesPerBurst() returns 1600. It was very hard to reproduce, but after several attempts the audio output was broken.

4) 920->1840. It was hard, but possible to reproduce.

5) 962! Just +2 from the default 960. For a couple of minutes I couldn't break the audio output, but then it happened.

Whenever I tried to return back to 960, the issue could be reproduced absolutely easily.

@philburk, I hope this information can be of any help. And if it looks like a hardware issue, we should close this.

**Comment by philburk** [comment]
>  Oboe can't start a LowLatency stream on my phone (I have previously sent the logcat example).
> So the opened stream has getPerformanceMode() == None. But it still uses the flowgraph

That may be an important clue. Huawei has disabled FAST tracks on many of their devices.

I may be able to reproduce this by letting Oboe set up the conversion flowgraph and then not requesting a FAST track from OpenSL ES.

**Comment by philburk** [comment]
I tried forcing performance mode NONE and cannot reproduce this.
Given that this is intermittent and has only been seen on a Huawei device running 8.0, It may be a device specific issue.
There is not much more we can do. So I am closing this. Sorry.

**Comment by serhii-k** [comment]
Yes, sure. Thank you very much for taking your time on investigating it.
